### PR TITLE
Fixed XmlAssert and added tests

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak @yishaigalatzer

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 1]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
         /// <param name="actualXml">Actual xml string.</param>
         public static void Equal(string expectedXml, string actualXml)
         {
-            var sortedExpectedXDoc = SortAttributes(XDocument.Parse(expectedXml));
-            var sortedActualXDoc = SortAttributes(XDocument.Parse(actualXml));
+			var sortedExpectedXDoc = XDocument.Parse(expectedXml).SortAttributes();
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: Something weird going on with these extra tab spaces...i will fix them later, but please check the other changes...

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 2]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
...
+
+			if (expected == null || actual == null)
+			{
+				return false;
+			}
+
+			return string.Equals(expected.Version, actual.Version, StringComparison.OrdinalIgnoreCase)
+				&& string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
+		}
+		private static XDocument SortAttributes(this XDocument doc)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: add blank line between these declarations

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 3]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
...
         {
-            var sortedExpectedXDoc = SortAttributes(XDocument.Parse(expectedXml));
-            var sortedActualXDoc = SortAttributes(XDocument.Parse(actualXml));
+			var sortedExpectedXDoc = XDocument.Parse(expectedXml).SortAttributes();
+			var sortedActualXDoc = XDocument.Parse(actualXml).SortAttributes();
 
-            bool areEqual = XNode.DeepEquals(sortedExpectedXDoc, sortedActualXDoc);
+			// Since XNode's DeepEquals does not check for presence of xml declaration,
+			// check it explicitly
+			bool areEqual = VerifyXmlDeclaration(sortedExpectedXDoc.Declaration, sortedActualXDoc.Declaration);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: `equalDeclarations`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 4]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
...
         {
             return new XDocument(
                     doc.Declaration,
-                    SortAttributes(doc.Root));
+					doc.Nodes().OfType<XText>().FirstOrDefault(),
+					doc.Root.SortAttributes());
         }
 
-        private static XElement SortAttributes(XElement element)
+		private static XElement SortAttributes(this XElement element)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: avoid an extension method here.  makes method harder to find.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 5]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
         /// <param name="actualXml">Actual xml string.</param>
         public static void Equal(string expectedXml, string actualXml)
         {
-            var sortedExpectedXDoc = SortAttributes(XDocument.Parse(expectedXml));
-            var sortedActualXDoc = SortAttributes(XDocument.Parse(actualXml));
+			var sortedExpectedXDoc = XDocument.Parse(expectedXml).SortAttributes();
+			var sortedActualXDoc = XDocument.Parse(actualXml).SortAttributes();
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: old names but fix them up to avoid abbreviations -> `...Document`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 1 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: My VS settings (on newest install) suddenly wanted to use tabs for everything. I had to change it.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: What's the point of this change?

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 4 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: :100: please don't randomly do this

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 6]
```diff
@@ -20,30 +22,62 @@ public static class XmlAssert
         /// <param name="actualXml">Actual xml string.</param>
         public static void Equal(string expectedXml, string actualXml)
         {
-            var sortedExpectedXDoc = SortAttributes(XDocument.Parse(expectedXml));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: add an early exit that returns `true` when `string.Equals(expectedXml, actualXml, StringComparison.Ordinal)`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNet.Mvc.Xml.Test
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: no `.Test` in test namespaces; always use namespace of the test subject - `Microsoft.AspNet.Mvc.Xml` in this case

Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: Yeah, i thought of doing it but here the XmlAssert is a utility present in this project itself.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 6 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: :+1:

**Comment by kichalla** [comment]
Updated. @rynowak @dougbu

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 8]
```diff
@@ -42,8 +70,24 @@ private static XElement SortAttributes(XElement element)
         {
             return new XElement(
                     element.Name,
+                    element.Nodes().OfType<XText>().FirstOrDefault(),
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: these uses of `FirstOrDefault()` will cause the equality checks to ignore additional text nodes.  need to include _all_ text nodes.  probably fine that you're ignoring comments and processing instructions but text nodes are (all) important.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 9]
```diff
@@ -20,21 +22,47 @@ public static class XmlAssert
...
+            }
+
+            return string.Equals(expected.Version, actual.Version, StringComparison.OrdinalIgnoreCase)
+                && string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
         }
 
         private static XDocument SortAttributes(XDocument doc)
         {
             return new XDocument(
                     doc.Declaration,
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: don't bother passing in the `Declaration`.  now handled elsewhere in this class.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 10]
```diff
@@ -42,8 +70,24 @@ private static XElement SortAttributes(XElement element)
...
             return new XElement(
                     element.Name,
+                    element.Nodes().OfType<XText>().FirstOrDefault(),
                     element.Attributes().OrderBy(a => a.Name.ToString()),
                     element.Elements().Select(child => SortAttributes(child)));
         }
+
+        // Since ToString() on an XDocument gives a formatted string making it difficult
+        // to view differences in strings when a test fails, here the formatting is disabled.
+        private static string GetRawXml(this XDocument xdocument)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: - again, do not make this an extension method.  call `GetRawXml(argument)` instead
- don't abbreviate; just call the parameter `document`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 11]
```diff
@@ -20,21 +22,47 @@ public static class XmlAssert
...
+            if (expected == null || actual == null)
+            {
+                return false;
+            }
+
+            return string.Equals(expected.Version, actual.Version, StringComparison.OrdinalIgnoreCase)
+                && string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
         }
 
         private static XDocument SortAttributes(XDocument doc)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: old name but don't abbreviate -> `document`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+	public class XmlAssertTest
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: looks like you're still using tabs; don't

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 13]
```diff
@@ -42,8 +70,24 @@ private static XElement SortAttributes(XElement element)
...
+        // to view differences in strings when a test fails, here the formatting is disabled.
+        private static string GetRawXml(this XDocument xdocument)
+        {
+            string xml = null;
+            var stream = new MemoryStream();
+            xdocument.Save(stream, SaveOptions.DisableFormatting);
+            stream.Position = 0;
+            using (var reader = new StreamReader(stream))
+            {
+                xml = reader.ReadToEnd();
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: no need for the `xml` variable.  just `return reader.ReadToEnd();` here.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+			"<?xml version=\"1.0\" encoding=\"UTF-16\"?><A></A>")]
+		public void Validates_XmlDeclaration(string input1, string input2)
+		{
+			Assert.Throws<EqualException>(() => XmlAssert.Equal(input1, input2));
+		}
+
+		[Fact]
+		public void Validates_XmlDeclaration_IgnoresCase()
+		{
+			// Arrange and Act
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: no Act here; it's `Act & Assert` on line 46

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: same for test methods below

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 6 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: I think we should not do this as we want to validate in case of invalid xmls being supplied(which would be missed if we had this check)..

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 9 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: I need this as I parse the xml strings to get the declarations and then do the comparison on them.

**Comment by kichalla** [comment]
Updated. @dougbu

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: are you saying the `expectedXml` value might be invalid?  why handle that case?

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 15]
```diff
@@ -20,30 +23,65 @@ public static class XmlAssert
...
+                    SortAttributes(document.Root));
         }
 
         private static XElement SortAttributes(XElement element)
         {
             return new XElement(
                     element.Name,
+                    element.Nodes().Where(node => node.NodeType != XmlNodeType.Element),
                     element.Attributes().OrderBy(a => a.Name.ToString()),
                     element.Elements().Select(child => SortAttributes(child)));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: the `NodeType != Element` ... `element.Elements()` pattern will reorder the nodes oddly.  won't affect the equality checks but will make it harder to read the `Message` of the `EqualException` when something is different.  e.g. compare `<doc>The <b>brown</b> dog</doc>` with `<doc>The <b>white</b> dog</doc>` and see errors about `<doc>The dog <b>brown</b></doc>` not matching `<doc>The dog <b>white</b></doc>`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+{
+    public class XmlAssertTest
+    {
+        [Theory]
+        [InlineData("<A>hello</A>", "<A>hey</A>")]
+        [InlineData("<A><B>hello world</B></A>", "<A><B>hello world!!</B></A>")]
+        [InlineData("<A>hello<B>world</B>hello</A>", "<A>hello<B>world!</B>hello</A>")]
+        public void ValidatesTextNodes(string input1, string input2)
+        {
+            Assert.Throws<EqualException>(() => XmlAssert.Equal(input1, input2));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: check `Message` here matches what you expect.  I expect it will mention `<A>hellohello<B>world</B></A>` and ``<A>hellohello<B>world!</B></A>`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 15 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: I suggest instead
``` c#
  private static XNode SortAttributes(XNode node)
  {
    XElement element = node as XElement;
    if (element == null)
    {
      return node;
    }

    return new XElement(
      element.Name,
      element.Attributes().OrderBy(...),
      element.Nodes().Select(child => SortAttributes(child)));
  }
```

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+        [Fact]
+        public void ValidatesAttributes_IgnoringOrder()
+        {
+            // Arrange
+            var input1 = "<A><B color=\"red\" size=\"medium\">hello world</B></A>";
+            var input2 = "<A><B size=\"Medium\" color=\"red\">hello world</B></A>";
+
+            // Act and Assert
+            Assert.Throws<EqualException>(() => XmlAssert.Equal(input1, input2));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: check `Message` matches what you expect here; should show sorted attributes

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 6 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: Since 'expectedXml` could be handcrafted when writing a test, it would be better to know that the xml is invalid.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 18]
```diff
@@ -20,30 +23,65 @@ public static class XmlAssert
...
+                return true;
+            }
+
+            if (expected == null || actual == null)
+            {
+                return false;
+            }
+
+            return string.Equals(expected.Version, actual.Version, StringComparison.OrdinalIgnoreCase)
+                && string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: nit: ignores `Standalone` property.  likely irrelevant for most comparisons but worth a comment in case someone starts testing some corner cases.

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 19]
```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    public class XmlAssertTest
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: add tests of
- differences in later text nodes e.g. `<a>hello<b>world</b>hello</a>` compared to `<a>hello<b>world</b>goodbye</a>`
- differences in later elements e.g. `<a><b>hello</b<b>world</b></a>` compared to `<a><b>hello</b<c>world</c></a>`, `<a><b>hello</b<b>goodbye</b></a>`, and `<a><b>hello</b<b attribute="value">world</b></a>`

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: that case involves two developer errors matching exactly and isn't worth worrying about.  but my comment was a nit and, by definition, doesn't require a fix if you disagree.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs (hunk 16 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: good point!

**Comment by kichalla** [comment]
@dougbu .Updated. Please take a look (I mistakenly squashed my previous commit, sorry about that.)

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 20]
```diff
@@ -20,30 +24,59 @@ public static class XmlAssert
...
+                return node;
+            }
+
             return new XElement(
-                    element.Name,
-                    element.Attributes().OrderBy(a => a.Name.ToString()),
-                    element.Elements().Select(child => SortAttributes(child)));
+              element.Name,
+              element.Attributes().OrderBy(a => a.Name.ToString()),
+              element.Nodes().Select(child => SortAttributes(child)));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: weird formatting.  why aren't arguments indented 4 spaces?

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs [hunk 21]
```diff
@@ -20,30 +24,59 @@ public static class XmlAssert
...
+                && string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
+        }
+
+        private static XDocument SortAttributes(XDocument document)
         {
             return new XDocument(
-                    doc.Declaration,
-                    SortAttributes(doc.Root));
+                    document.Declaration,
+                    SortAttributes(document.Root));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs]: weird formatting

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+        [InlineData("<A></A>", "<A></A>")]
+        [InlineData("<A>hello<B>world</B>hello</A>", "<A>hello<B>world</B>hello</A>")]
+        [InlineData("<A><!-- comment1 --><B></B></A>", "<A><!-- comment1 --><B></B></A>")]
+        [InlineData("<A/>", "<A/>")]
+        [InlineData("<A><![CDATA[<greeting></greeting>]]></A>", "<A><![CDATA[<greeting></greeting>]]></A>")]
+        [InlineData("<a>hello<b>world</b>hello</a>", "<a>hello<b>world</b>hello</a>")]
+        [InlineData(
+            "<a>hello<b color=\"red\" siz=\"medium\">world</b>hello</a>",
+            "<a>hello<b siz=\"medium\" color=\"red\">world</b>hello</a>")]
+        public void DoesNotFail_OnEmptyTextNodes(string input1, string input2)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: rename to `Validates_WithEmptyElements` to avoid mentioning an impossible case; XML parsers don't create empty text nodes

File: test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs (hunk 19 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssertTest.cs]: don't see tests of
- `<a>hello<b>world</b>hello</a>` compared to `<a>hello<b>world</b>goodbye</a>` (different 2nd text node)
- `<a><b>hello</b<b>world</b></a>` compared to `<a><b>hello</b<b>goodbye</b></a>` (different text in second element)

also, realized you're using "Validates" for both tests that `throw` and don't.  suggest
- `Throws_WithMismatchedTextNodes`
- `ReturnsSuccessfully_WithEmptyElements`
- `Throws_WithMismatchedXmlDeclarations`
- `ReturnsSuccessfully_WithMatchingXmlDeclaration`
- `ReturnsSuccessfully_WithMatchingXmlDeclaration_IgnoringCase`
- `ReturnsSuccessfully_WithMatchingContent_IgnoringAttributeOrder`
- `Throws_WithMismatchedAttributeValues_ReorderingAttributes`

or use the scenario_expectation order e.g. `MismatchedAttributeValues_Throws_ReorderingAttributes`

**Comment by dougbu** [comment]
:watch:

**Comment by kichalla** [comment]
Updated. /cc: @dougbu

**Comment by dougbu** [comment]
:shipit:
